sidecar/internal/volume-condition/reporter.go

```diff
@@ -0,0 +1,178 @@
+	er, err := NewEventRecorder(client, hostname)
+	_ = err
+	recorders = append(recorders, er)
+
+	return &volumeConditionReporter{
+		client:         client,
+		recorders:      recorders,
+		driver:         drv,
+		localNode:      n,
+		conditionCache: make(map[string]volume.VolumeCondition, 0),
+	}, nil
+}
+
+func (cvr *volumeConditionReporter) Run(ctx context.Context, interval time.Duration) error {
+	running := time.Tick(interval)
+	if running == nil {
+		return fmt.Errorf("interval %v is invalid", interval)
+	}
+
+	for range running {
+		volumes, err := cvr.localNode.ListVolumes(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to list volumes: %w", err)
+		}
+
+		for _, v := range volumes {
+			if v.GetDriver() != cvr.driver.GetDrivername() {
+				continue
+			}
+
+			vc, err := cvr.driver.GetVolumeCondition(v)
+			if err != nil {
+				klog.Errorf("failed to check if %q is healthy: %v", v.GetVolumeID(), err)
+				continue
+			}
+
+			if !cvr.isUpdatedVolumeCondition(v.GetVolumeID(), vc) {
+				// skip recording if there is no update
+				continue
+			}
+
+			cvr.recordVolumeCondition(ctx, v.GetVolumeID(), vc)
+		}
+
+		cvr.pruneConditionCache(volumes)
+
+		if cvr.shouldStop {
+			break
+		}
+	}
+
+	return nil
+}
+
+func (cvr *volumeConditionReporter) Stop() {
+	cvr.shouldStop = true
+}
+
+func (cvr *volumeConditionReporter) isUpdatedVolumeCondition(volumeID string, vc volume.VolumeCondition) bool {
+	lastCondition := cvr.conditionCache[volumeID]
+	if lastCondition == nil || (lastCondition.IsHealthy() != vc.IsHealthy() || lastCondition.GetMessage() != vc.GetMessage()) {
+		cvr.conditionCache[volumeID] = vc
+		return true
+	}
+
+	return false
+}
+
+// pruneConditionCache removed inactive volume-handles from the cvr.conditionCache. This is done by creating a
+// new cache, and have the old cvr.conditionCache get garbage collected.
+func (cvr *volumeConditionReporter) pruneConditionCache(volumes []volume.Volume) {
+	newConditionCache := make(map[string]volume.VolumeCondition, len(volumes))
+	for _, v := range volumes {
+		volumeID := v.GetVolumeID()
+		newConditionCache[volumeID] = cvr.conditionCache[volumeID]
+	}
+
+	cvr.conditionCache = newConditionCache
+}
+
+func (cvr *volumeConditionReporter) recordVolumeCondition(ctx context.Context, volumeID string, vc volume.VolumeCondition) {
+	pvName, err := platform.GetPlatform().ResolvePersistentVolumeName(cvr.driver.GetDrivername(), volumeID)
+	if err != nil {
+		klog.Errorf("failed to resolve persistent volume name: %v", err)
+		return
+	}
+
+	pv, err := cvr.client.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("failed to get persistent volume %q: %v", pvName, err)
+		return
+	}
+
+	for _, recorder := range cvr.recorders {
+		err = recorder.Record(ctx, pv, vc)
+		if err != nil {
+			klog.Warningf(
+				"%T failed to record volume condition for persistent volume %q: %v",
+				recorder,
+				pv.Name,
+				err,
+			)
+		}
+	}
```

Comment:
dont we need to record this at the PVC level as that is the user created object? the PV is admin created object and with some restriction users might not have access to PV objects as its cluster scoped?